File: paddle/ir/core/parser/lexer.cc
// Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "paddle/ir/core/parser/lexer.h"
using std::string;

Token Lexer::GetToken(LexSegment seg) {
  SkipWhitespace();
  if (auto token = LexIdentifer(seg)) {
    return *token;
  } else if (auto token = LexNumberOrArraow()) {
    return *token;
  } else if (auto token = LexEndTagOrNullVal(seg)) {
    return *token;
  } else if (auto token = LexValueId()) {
    return *token;
  } else if (auto token = LexOpName()) {
    return *token;
  } else if (auto token = LexEOF()) {
    return *token;
  } else {
    return Token{"Error", NULL_};
  }
}

char Lexer::GetChar() {
  char c = is.get();
  if (c == '\n') {
    line++;
    column = 1;
  } else {
    column++;
  }
  return c;
}

size_t Lexer::GetColumn() { return column; }

size_t Lexer::GetLine() { return line; }

void Lexer::SkipWhitespace() {
  while (IsSpace(is.peek())) {
    GetChar();
  }
}

Token* Lexer::LexIdentifer(LexSegment seg) {
  if ((!isalpha(is.peek()) && is.peek() != '_') || IsEndTag(is.peek(), seg)) {
    return nullptr;
  }
  string token_identifier = "";
  while (isalnum(is.peek()) || is.peek() == '_' || is.peek() == '.') {
    token_identifier += GetChar();
  }
  return new Token{token_identifier, IDENTIFER};
}

Token* Lexer::LexNumberOrArraow() {
  if (!isdigit(is.peek()) && is.peek() != '-') {
    return nullptr;
  }

  string token_digit = "";
  token_digit += GetChar();

  if (token_digit[0] == '-' && is.peek() == '>') {
    GetChar();
    return new Token{"->", ARRAOW};
  }
  while (isdigit(is.peek())) {
    token_digit += GetChar();
  }
  if (is.peek() == '.') {
    token_digit += GetChar();
    while (isdigit(is.peek())) {
      token_digit += GetChar();
    }
  }
  if (is.peek() == 'e') {
    token_digit += GetChar();
    if (is.peek() == '+' || is.peek() == '-') {
      token_digit += GetChar();
    }
    while (isdigit(is.peek())) {
      token_digit += GetChar();
    }
    return new Token{token_digit, SDIGIT};
  }
  return new Token{token_digit, DIGIT};
}

Token* Lexer::LexEndTagOrNullVal(LexSegment seg) {
  if (!IsEndTag(is.peek(), seg)) {
    return nullptr;
  }
  string token_end = "";
  token_end += GetChar();
  if ((token_end[0] == '<' && (is.peek() != '<' && is.peek() != '#')) ||
      token_end[0] != '<') {
    return new Token{token_end, ENDTAG};
  }
  if (is.peek() == '<') {
    string token_null_val = "";
    GetChar();
    while (is.peek() != '>') {
      token_null_val += GetChar();
    }
    GetChar();
    GetChar();
    return new Token{"<<" + token_null_val + ">>", NULL_};
  } else {
    string token_attrnull = "";
    while (is.peek() != '>') {
      token_attrnull += GetChar();
    }
    GetChar();
    return new Token{"<" + token_attrnull + ">", NULL_};
  }
}

Token* Lexer::LexValueId() {
  if (is.peek() != '%') {
    return nullptr;
  }
  string token_valueid = "";
  token_valueid += GetChar();

  while (isdigit(is.peek())) {
    token_valueid += GetChar();
  }
  return new Token{token_valueid, VALUEID};
}

Token* Lexer::LexEOF() {
  if (is.peek() == EOF) {
    return new Token{"LEX_DOWN", EOF_};
  } else {
    return nullptr;
  }
}

Token* Lexer::LexOpName() {
  if (is.peek() != '"') {
    return nullptr;
  }
  GetChar();
  string token_opname = "";
  while (is.peek() != '"') {
    token_opname += GetChar();
  }
  GetChar();
  return new Token{token_opname, OPNAME};
}

bool Lexer::IsSpace(char c) {
  return c == ' ' || c == '\n' || c == '\t' || c == '\f';
}

bool Lexer::IsEndTag(char c, LexSegment seg) {
  if (seg == parseFunctionType) {
    return c == '{' || c == '}' || c == '(' || c == ')' || c == ':' ||
           c == '>' || c == ',' || c == ']' || c == '[' || c == '+' ||
           c == '=' || c == 'x' || c == '<';
  } else {
    return c == '{' || c == '}' || c == '(' || c == ')' || c == ':' ||
           c == '>' || c == ',' || c == ']' || c == '[' || c == '+' ||
           c == '=' || c == '<';
  }
}

File: python/paddle/distributed/flex_checkpoint/aoa/lexer.py
# Copyright (c) 2025 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import re
from enum import Enum, auto


class Token:
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __repr__(self):
        return f"Token({self.type}, {self.value!r})"


class TokenType(Enum):
    IDENTIFIER = auto()
    NUMBER = auto()
    COLON = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    COMMA = auto()
    RARROW = auto()
    STRING = auto()
    EQUAL = auto()
    NEWLINE = auto()
    EOF = auto()


class Lexer:
    token_specification = [
        ('RARROW', r'->'),
        ('EQUAL', r'='),
        ('COLON', r':'),
        ('LBRACKET', r'\['),
        ('RBRACKET', r'\]'),
        ('COMMA', r','),
        ('NUMBER', r'\d+'),
        ('STRING', r'"[^"]*"|\'[^\']*\''),
        ('IDENTIFIER', r'[A-Za-z_][A-Za-z\.\$\_\*\d\^T]*'),
        ('SKIP', r'[ \t]+'),
        ('NEWLINE', r'[\r\n]+'),
        ('MISMATCH', r'.'),
    ]

    def __init__(self, context):
        from .macros import macro_registry

        self.macros = [list(d.values())[1] for d in macro_registry.macros]
        self.get_token = re.compile(
            '|'.join(
                f'(?P<{name}>{regex})'
                for name, regex in self.token_specification
            )
        ).match
        self.context = context

    def tokenize(self, text):
        pos = 0
        mo = self.get_token(text, pos)
        tokens = []
        if not text.endswith('\n'):
            text += '\n'
        while mo is not None:
            kind = mo.lastgroup
            value = mo.group()
            if kind == 'SKIP':
                pass
            elif kind == 'MISMATCH':
                raise RuntimeError(
                    f'Unexpected character {value!r} at position {pos}'
                )
            else:
                tokens.append(Token(TokenType[kind], value))
            pos = mo.end()
            mo = self.get_token(text, pos)
        return tokens

    def apply_macro(self, expression, macro):
        if isinstance(expression, str):
            expression = [expression]
        new_expression = []
        for expr in expression:
            results = macro(self.tokenize(expr), expr, self.context)
            if isinstance(results, str):
                new_expression.append(results)
            else:
                new_expression.extend(results)
        return new_expression

    def apply_single_macro_to_all(self, expressions, macro):
        new_expressions = []
        for expr in expressions:
            results = macro(self.tokenize(expr), expr, self.context)
            if isinstance(results, str):
                new_expressions.append(results)
            else:
                new_expressions.extend(results)
        return new_expressions

    def all_tokens(self, expressions):
        current_expressions = expressions
        for macro in self.macros:
            current_expressions = self.apply_single_macro_to_all(
                current_expressions, macro
            )

        tokens = []
        for expr in current_expressions:
            tokens.extend(self.tokenize(expr))
        return tokens
